From 1d729fc37c25a95cd52ebaa9ca04658903acd85b Mon Sep 17 00:00:00 2001
From: Michael Zhao <michael.zhao@arm.com>
Date: Tue, 7 Nov 2023 15:47:00 +0800
Subject: [PATCH] crypto: Support psa_destroy_key in wrapper

Add the entry for `psa_destroy_key` in the wrapper. Now the wrapper
entry support RSS communication driver only.

Upstream-Status: Pending
Signed-off-by: Michael Zhao <michael.zhao@arm.com>
Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>
Signed-off-by: Divin Raj <divin.raj@arm.com>
---
 library/psa_crypto.c                          |  9 ++++++
 library/psa_crypto_driver_wrappers.c          | 30 ++++++++++++++++++
 library/psa_crypto_driver_wrappers.h          |  4 +++
 .../psa_crypto_driver_wrappers.c.jinja        | 31 +++++++++++++++++++
 4 files changed, 74 insertions(+)

diff --git a/library/psa_crypto.c b/library/psa_crypto.c
index 13749f783..320e547cc 100644
--- a/library/psa_crypto.c
+++ b/library/psa_crypto.c
@@ -1183,6 +1183,15 @@ psa_status_t psa_destroy_key(mbedtls_svc_key_id_t key)
     }
 #endif /* MBEDTLS_PSA_CRYPTO_SE_C */
 
+    psa_key_attributes_t attributes = {
+        .core = slot->attr
+    };
+    status = psa_driver_wrapper_destroy_key(&attributes,
+                                            slot->key.data, slot->key.bytes);
+   if( overall_status == PSA_SUCCESS )
+        overall_status = status;
+
+
 exit:
     status = psa_wipe_key_slot(slot);
     /* Prioritize CORRUPTION_DETECTED from wiping over a storage error */
diff --git a/library/psa_crypto_driver_wrappers.c b/library/psa_crypto_driver_wrappers.c
index f48452b0a..b20d87c82 100644
--- a/library/psa_crypto_driver_wrappers.c
+++ b/library/psa_crypto_driver_wrappers.c
@@ -1225,6 +1225,36 @@ psa_status_t psa_driver_wrapper_export_public_key(
 
 }
 
+psa_status_t psa_driver_wrapper_destroy_key(
+    const psa_key_attributes_t *attributes,
+    const uint8_t *key_buffer, size_t key_buffer_size)
+{
+
+    psa_status_t status = PSA_SUCCESS;
+    psa_key_location_t location = PSA_KEY_LIFETIME_GET_LOCATION(
+                                      psa_get_key_lifetime( attributes ) );
+
+    switch( location )
+    {
+#if defined(PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT)
+#if defined(MBEDTLS_RSS_EXAMPLE_DRIVER_ENABLED)
+        case PSA_KEY_LOCATION_RSS:
+            status = rss_opaque_destroy_key( attributes,
+                                             key_buffer,
+                                             key_buffer_size );
+            return( status );
+
+#endif /* MBEDTLS_RSS_EXAMPLE_DRIVER_ENABLED */
+#endif /* PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT */
+
+        case PSA_KEY_LOCATION_LOCAL_STORAGE:
+        default:
+            /* Key is declared with a lifetime not known to us */
+            return( status );
+    }
+
+}
+
 psa_status_t psa_driver_wrapper_get_builtin_key(
     psa_drv_slot_number_t slot_number,
     psa_key_attributes_t *attributes,
diff --git a/library/psa_crypto_driver_wrappers.h b/library/psa_crypto_driver_wrappers.h
index 7d672d641..21952c044 100644
--- a/library/psa_crypto_driver_wrappers.h
+++ b/library/psa_crypto_driver_wrappers.h
@@ -123,6 +123,10 @@ psa_status_t psa_driver_wrapper_export_public_key(
     const uint8_t *key_buffer, size_t key_buffer_size,
     uint8_t *data, size_t data_size, size_t *data_length);
 
+psa_status_t psa_driver_wrapper_destroy_key(
+    const psa_key_attributes_t *attributes,
+    const uint8_t *key_buffer, size_t key_buffer_size);
+
 psa_status_t psa_driver_wrapper_get_key_buffer_size(
     const psa_key_attributes_t *attributes,
     size_t *key_buffer_size);
diff --git a/scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja b/scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja
index 08da774d3..8a9bbd68d 100644
--- a/scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja
+++ b/scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja
@@ -1783,6 +1783,37 @@ psa_status_t psa_driver_wrapper_hash_abort(
     }
 }
 
+psa_status_t psa_driver_wrapper_destroy_key(
+    const psa_key_attributes_t *attributes,
+    const uint8_t *key_buffer, size_t key_buffer_size)
+{
+
+    (void) key_buffer;
+    (void) key_buffer_size;
+    psa_status_t status = PSA_SUCCESS;
+    psa_key_location_t location = PSA_KEY_LIFETIME_GET_LOCATION(
+                                      psa_get_key_lifetime( attributes ) );
+
+    switch( location )
+    {
+#if defined(PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT)
+#if defined(MBEDTLS_RSS_EXAMPLE_DRIVER_ENABLED)
+        case PSA_KEY_LOCATION_RSS:
+            status = rss_opaque_destroy_key(key_buffer,
+                                             key_buffer_size );
+            return( status );
+
+#endif /* MBEDTLS_RSS_EXAMPLE_DRIVER_ENABLED */
+#endif /* PSA_CRYPTO_ACCELERATOR_DRIVER_PRESENT */
+
+        case PSA_KEY_LOCATION_LOCAL_STORAGE:
+        default:
+            /* Key is declared with a lifetime not known to us */
+            return( status );
+    }
+
+}
+
 psa_status_t psa_driver_wrapper_aead_encrypt(
     const psa_key_attributes_t *attributes,
     const uint8_t *key_buffer, size_t key_buffer_size,
-- 
2.25.1

